In selector check, prefix of reference must match import qualifier

som-snytt opened the pull request:
This PR changes the `CheckUnused` phase to rely on the `MiniPhase` API (instead of custom traversal). That improves fidelity to `Context` (instead of approximate scoping).

The phase should work seamlessly with subsequent linting phases (currently, `CheckShadowed`).

It is a goal of the PR to eliminate false reports. It is also a goal not to regress previous work on efficiency.

A remaining limitation of the current approach is that contexts don't provide a nesting level. Practically, this means that for a wildcard import nested below a higher precedence named import, the wildcard is deemed "unused". (A more general tool for "managing" or "formatting" imports could do more to pick a preferred scope.)

This PR adds `-Wunused:patvars`, as forward-ported from Scala 2: it relies on attachments for some details about desugaring, but otherwise uses positions (where only the original patvar has a non-synthetic position).

As in Scala 2, it does not warn about patvars with the "canonical" name of a case class element (this is complicated by type tests and the quotes API); other exclusions are to be ported, such as "name derived from the match selector".

Support is added for `-Wconf:origin=full.path.selector`, as in Scala 2. That allows, for example:
```
-Wconf:origin=scala.util.chaining.given:s
```
to exclude certain blessed imports from warnings, or to work around false positives (should they arise).

Support is added to `-rewrite` unused imports. There are no options to "format"; instead, textual deletions preserve existing formatting, except that blank lines are removed and braces removed when there is only one selector.

Notable fixes are to support `compiletime` and `inline`; there are more fixes to pursue in this area.

The commits are not organized around these changes; ~commits are preserved here just for comparison to previous art, so that useful existing behaviors do not regress~.

Fixes #19657
Fixes https://github.com/scala/scala3/issues/20520
Fixes https://github.com/scala/scala3/issues/19998
Fixes https://github.com/scala/scala3/issues/18313
Fixes #17371
Fixes #18708
Fixes https://github.com/scala/scala3/issues/21917
Fixes https://github.com/scala/scala3/issues/21420
Fixes https://github.com/scala/scala3/issues/20951
Fixes https://github.com/scala/scala3/issues/19252
Fixes https://github.com/scala/scala3/issues/18289
Fixes https://github.com/scala/scala3/issues/17667
Fixes https://github.com/scala/scala3/issues/17252
Fixes https://github.com/scala/scala3/issues/21807
Fixes https://github.com/scala/scala3/issues/17753
Fixes https://github.com/scala/scala3/issues/17318
Fixes https://github.com/scala/scala3/issues/18564
Fixes https://github.com/scala/scala3/issues/22376
Fixes https://github.com/scala/scala3/issues/21525

som-snytt commented:
Further tweaks forthcoming.

Gedochao commented:
@som-snytt how much work do you think it still requires? 
It seems we might want to backport the fix for #20860  to 3.5.0 at the very least... 
do you need any help?

sjrd commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -352,46 +353,46 @@ object CheckUnused:
      */
-    private val usedInScope = MutStack(MutSet[(Symbol, Option[Name], Boolean)]())
-    private val usedInPosition = MutMap.empty[Name, MutSet[Symbol]]
+    private val usedInScope = Stack(mut.Set.empty[(Symbol, Option[Name], Type, Boolean)])
```
This is suspicious. `Type`s don't have a meaningful `==` test in general. Why is it OK to use a `Type` in a `Set` here?

sjrd commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -714,12 +713,13 @@ object CheckUnused:
             // if there is an explicit name, it must match
             false
-          else
+          else selData.qualTpe =:= prefix && (
```
Let's put this on a dedicated line ;)
```suggestion
          else
            selData.qualTpe =:= prefix && (
```

sjrd commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -833,7 +834,7 @@ object CheckUnused:
         case _ => Other
 
-    final class ImportSelectorData(val qualTpe: Type, val selector: ImportSelector):
+    final case class ImportSelectorData(val qualTpe: Type, val selector: ImportSelector):
```
This seems unnecessary.

som-snytt replied:
sadly I needed string debug

sjrd reviewed (changes requested):
The first commit is a bit suspicious, and might be the cause of the CI failures.

The second commit looks good and might pass more easily if submitted as an independent PR.

som-snytt commented:
I'll submit commit 2 separately, as I was about to do except it's one line.

I'll follow up the fix for commit 1 after the American holiday.

som-snytt commented:
The previous test fails were because the new prefix check (in `isInImport`) doesn't apply when `isDerived` and also for top-level defs where the prefix becomes a package object (not sure if that `=:=` test should work, or if there's a convenient idiom, or if the check is even useful). The derived check uses dealiased types, not sure if a better mechanism is warranted.

I'll clean up and also look for improvements.

som-snytt commented:
Rebased and split out commits for easier review.

This includes adding the prefix type to `Usage`; simplifying gathering the warnings in `getUnused`, avoiding intermediate collections and extra sorting; mild refactors for readability.

This is the July commit and not the August rewrite, which fixes scope bugs by leveraging normal miniphase traversal and avoiding the special traverser; which allows putting `CheckUnused` in a single megaphase with `CheckShadowing`, so that is worth returning to. I'll do that if this PR of limited scope receives a review.

Example extra test fixed in August:
```
object Constants:
  val i = 42
class `scope of super`:
  import Constants.i // bad warn
  class C(x: Int):
    def y = x
  class D extends C(i):
    import Constants.* // does not resolve i in C(i)
    def m = i
```
As the comment reminds me, it incorrectly detects the superconstructor ref as resolved by the nested import.

Also note that the inner wildcard is not ambiguous because both imports resolve to the same symbol. But a further improvement would be to warn that the inner import is, if not unused, then spurious.

som-snytt commented:
Cherry-picked some old commits.

As a reminder to future self, the mega phase was broken because CheckShadowing sees the nested X as shadowing.
```
class F[X, M[N[X]]]
```
The fix [sic] is to ignore everything nested because it ignores constructor-owned X and M. [Previous attempt had been to short-circuit transformDeep so CheckShadowing doesn't see subtrees of "other" trees.]

The fix [also sic] to superclass context noted in previous comment is also novel: since all the "context" approximation is for the purpose of import tracking, it doesn't need to capture all the subtlety of a true super class constructor context (with class parameters in scope). Instead, just detect that a reference occurred in a parent tree, and then kick it upstairs in popScope.

The mechanism for tracking `derives` is entirely reworked.

The "inner traverser" is removed in favor of matching "other" trees and transforming their parts. I see there was some long discussion about this. 

The miniphase should be rewritten to leverage existing Context and Scope instead of custom structures. The problem of tracking just symbols is simpler (as in Scala 2), but for import tracking, it should use ImportContext etc. (The other PR for import tracking followed Scala 2 and let the compiler record lookups as they happen.) The motivation for proper contexts is so that a "lint framework" can rely on it, instead of custom traversals. For this phase, top-down may be more efficient than bubbling up references until they find a definition or import. Per context, a reference need be looked up only once, and then only if there are outstanding unused definitions or imports.

som-snytt commented:
`warn/unused-privates` has `NO warn` comments to revisit.

som-snytt commented:
> I'll follow up the fix for commit 1 after the American holiday.

That is Thanksgiving Day shortly after the national election, of course.

som-snytt commented:
I see I was in the middle of adding commits a month ago and did not implement the speculation from my previous comment.

My brain glazes over when I look at this PR. Also I have to re-read "getting started as a contributor" to understand "testing your changes". But I see this PR includes showing vulpix results more nicely.

som-snytt commented:
There are `pos` tests that check for "no warnings" using `-Werror` as in Scala 2 partest, but should be under `warn` which will check that is produces zero warnings.

som-snytt commented:
I didn't get around yet to handling constructor contexts when I worked around superconstructor contexts. https://github.com/scala/scala3/issues/21917

Note to self: don't neglect secondary constructors, which are lexically nested but typechecked like primary ctor in outer context.

som-snytt commented:
That link is a reminder to add `origin` to diagnostics.

som-snytt commented:
squashed because couldn't be bothered to rebase the conflict. I saved a branch in case.

Still need to fix the unpickle test failure. Then I will book the progress to avoid further conflicts. More accommodations are needed for compiletime.

som-snytt commented:
This breaks the unpickler test, or rather this is what unbreaks it:
```
--- a/compiler/src/dotty/tools/dotc/Compiler.scala
+++ b/compiler/src/dotty/tools/dotc/Compiler.scala
@@ -51,7 +51,7 @@ class Compiler {
     List(new Staging) ::            // Check staging levels and heal staged types
     List(new Splicing) ::           // Replace level 1 splices with holes
     List(new PickleQuotes) ::       // Turn quoted trees into explicit run-time data structures
-    List(CheckUnused.PostInlining()) ::  // Check for unused elements
+    List(new CheckUnused.PostInlining) ::  // Check for unused elements
     Nil
```
I don't have bandwidth to investigate why.

som-snytt commented:
@sjrd This is "ready for review."

`resolveUsage` is the logic for deciding whether a name was obtained from an import or definition. (That includes the fix in the PR title.)

It walks `Context` owners like a name look-up, but attempts some caching at certain contexts. That is for correctness and hopefully efficiency (but I haven't tested performance yet).

`RefInfos` contains the import, definitions, and references that are collected. It omits the stacks for scoping from the previous implementation.

It attempts to track nested `Inlined` call site positions: resolveUsage happens if the current tree is at a call pos, or if it looks like `summonInline` (by its span of zero extent). It also counts `inline def` to avoid warning there; a new option (to be done) will enable linting inline defs. This is similar to Scala 2, where macro expansions "consume" elements but don't introduce new warnings; but an option allows linting before or after expansion.

There are various tiresome mechanisms which are expected to just work without tweaking options.

Another principle is to minimize surface area with other components; this is restricted to a few attachments.

A few minor clean-ups remain, including ensuring that `compiletime` doesn't warn (because information is elided early). I speculated that it might be useful to run again after erasure. Avoiding false positives is a high priority, but also inline and compiletime (or let's just say language features) should have linting support. I haven't verified that lints are amenable to suppression; there are tickets for nowarn; but suppression is deemed a last resort.

There are 100 LOC for deleting unused imports which should be moved to a cupboard, as it is supplementary, to be brought out on special occasions; it's a port from Scala 2.

som-snytt commented on library/src/scala/compiletime/package.scala:
Thanks, I noticed it: I had touched then reverted the file but haven't fixed this permissions glitch yet. I'd like to figure out what lapse of gitfu caused it.

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+      transformAllDeep(result)
+    case _: InferredTypeTree => // do nothing
+    case _: Export => // nothing to do
+    case _ if tree.isType =>
+      //println(s"OTHER TYPE ${tree.getClass} ${tree.show}")
+    case _ =>
+      //println(s"OTHER ${tree.getClass} ${tree.show}")
```
putting them back under a line comment *was* the clean-up!

som-snytt commented on tests/warn/i15503f.scala:
```diff
@@ -6,9 +6,46 @@ val default_int = 1
+    def unapply(x: Expr[Option[T]])(using Quotes) = x match
+      case '{ Option[T](${Expr(y)}) } => Some(Option(y))
+      case '{ None } => Some(None)
+      //case '{ ${Expr(opt)} : Some[T] } => Some(opt) // make Type param unused after typer
```
It's the same as the previous case, but the parameter `using Type` is unused, but does not warn under `-Wunused:implicits`. Now I'm not sure why. It's not named and `Type` has no declared members.

som-snytt replied:
Apparently something to do with quote pickling, not printed by -Vprint. Is there a tool for showing trees that I don't know about? How can I reason about data structures that are invisible?

som-snytt replied:
To follow up, the TODO was that I wanted to show the implicit arg unused, but didn't know enough about quoting to do it; in addition, I lacked tooling just to show me the `Ident` tree in question.

som-snytt commented on tests/warn/i15503kb/power.scala:
```diff
@@ -0,0 +1,15 @@
+
+object Power:
+  import scala.concurrent.* // warn [taps mic]
```
Make sure it's warning. Inside these tests, if it's testing that elements do *not* warn, then you'd never know if warnings were turned off (for whatever feature). In retrospect, I'm leaning toward preferring a fixed corpus which is run under various settings: then all the code is tried under all the options, in some combination. (The fixed corpus need not be one huge file, but a test group.)

som-snytt replied:
I've begun (recently) to add `-Werror` in tests/warn to signal that no warnings are expected. That assists clarity. If there is a warning, though, you don't get a nice vulpix explanation.

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+    ) || (
+         sel.isWildcard && sel.isGiven
+      && imp.expr.tpe.allMembers.exists(_.symbol.isCanEqual)
+      || imp.expr.tpe.member(sel.name.toTermName).hasAltWith(_.symbol.isCanEqual)
```
I'd also like to delete that whole thing: I don't think `strict no implicit warn` is useful if the check is working correctly (as it ought to be), and also I'm not sure `CanEqual` needs special treatment any more.

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
-      (imp.expr.tpe.member(sel.name.toTermName).alternatives
-        .exists(p => p.symbol.isOneOf(GivenOrImplicit) && p.symbol.typeRef.baseClasses.exists(_.derivesFrom(defn.CanEqualClass))))
+  extension (imp: Import)
+    /** Is it the first import clause in a statement? `a.x` in `import a.x, b,{y, z}` */
```
!

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+            // with aliases into requiredModule("dotty.tools.dotc.ast.tpd")
+            val implName = s"dotty.tools.dotc.ast.Trees$$${unapplied.name}"
+            try
+              import scala.language.unsafeNulls
+              val clz = Class.forName(implName)
+              val ok = clz.getConstructors.head.getParameters.map(p => termName(p.getName)).toList.init
```
Yes, I didn't want to read class files just for this "cheapy" attempt to get some param names for a warning. This is so `case Ident(name) =>` doesn't warn because `name` is the name. I'll look into improving it. The issue is the decoupling in the quotes api.

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+        case Typed(rhs, _) => isUnconsuming(rhs)
+        case _ => false
+
+  def absolveVariableBindings(ok: List[Name], args: List[Tree]): Unit =
```
Then the mathematicians must not say "absolute value" or the chemist "absolute alcohol" or the Swedish distiller "Absolut". The condition of absolution is to be freed from impairment, also in property rights. I feel more strongly about `Bind`, as this series of meanings relies on a primary binding from which one must be released (as from any obligation or duty, from which one may be absolved) in order to be fully realized. Excuse is just the dual of accuse. I'll go with "allow", allouer, laudare, which I didn't know is such a pleasant word, and with that end this lark or alouette.

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+  // The RHS of a def is too trivial to warn about unused params, e.g. def f(x: Int) = ???
+  def isUnconsuming(rhs: Tree)(using Context): Boolean =
+        rhs.symbol == defn.Predef_undefined
+     || rhs.tpe =:= defn.NothingType
```
`compiletime.error`. If it produces no value, then the parameter is an input to nothing (though it could appear in an exception message, etc).

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+        if tree.symbol ne NoSymbol then
+          defs.addOne((tree.symbol, tree.srcPos))
+
+    val inlined = Stack.empty[SrcPos] // enclosing call.srcPos of inlined code
+    var inliners = 0 // depth of inline def
```
one is you saw an `Inlined`, the other is you saw `inline def`. For `Inlined`, avoid an inlined reference "resolving" an import (there was an existing test https://github.com/scala/scala3/blob/3.6.3/tests/warn/i15503a.scala#L71) and conversely in `inline def` an import might be resolved only after inlining a summonInline. There's a new ticket and test for that. However, all this is probably brittle and needs further work. I didn't want to exonerate or excuse or allow just everything. Or exonerate might be a good word, as the sense of exculpate is secondary. It means relieve as in relieve one's bladder or bowels.

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+      case tree: NamedDefTree =>
+        if (tree.symbol ne NoSymbol) && !tree.name.isWildcard then
+          defs.addOne((tree.symbol, tree.namePos))
+      case _ =>
+        if tree.symbol ne NoSymbol then
+          defs.addOne((tree.symbol, tree.srcPos))
```
I intended to answer that question & will see what pops out here.

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+    var candidate: Context = NoContext
+    var cachePoint: Context = NoContext // last context with Resolved cache
+    var importer: ImportSelector | Null = null // non-null for import context
+    var precedence = Int.MaxValue // of current resolution; lower is higher
```
you shoot down my best turns of phrase. But it should be more symbolic than 1..4 so I'll improve that.

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
-          traverseChildren(tp)
+  def resolveUsage(sym: Symbol, name: Name, prefix: Type)(using Context): Unit =
+    def matchingSelector(info: ImportInfo): ImportSelector | Null =
+      val qtpe = info.site //info.qualifier.tpe.nn
```
ah I see qualifier is empty for root imports, which are consulted now. The reason for drilling up into root imports is to cache a result (`Object`). (Though I suppose it could stop once it reached a root import with the same result.)

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+            candidate = cur
+            importer = null
+            done = true
+          else if precedence > 4 then
+            precedence = 4
+            candidate = cur
```
defs in other source files are lower precedence than imports

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+    val ctxs = ctx.outersIterator
+    while !done && ctxs.hasNext do
+      val cur = ctxs.next()
+      if cur.owner eq sym then
+        foundEnclosing = true
+        done = true
```
I had already used one return & I'm sure tpolecat's ears were burning, but I will persevere and add another.

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
-        loopOnNormalizedPrefixes(tree.typeOpt.normalizedPrefix, depth = 0)
-        ud.registerUsed(tree.symbol, Some(tree.name))
-      }
+      // resolve if inlined at the position of the call, or is zero extent summon
```
That is the normal case, the tree is not an inline expansion. But now I think most Scala 3 code is inlined. Also this reverse-engineering from the synthetic position was just an observation; I'm not sure this is robust.

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+      resolveUsage(tree.tpe.classSymbol, tree.name, tree.tpe.importPrefix.skipPackageObject)
+    tree
+
+  // import x.y; x may be rewritten x.y, also import x.z as y
```
that is https://github.com/scala/scala3/blob/3.6.3/compiler/src/dotty/tools/dotc/typer/Typer.scala#L586 except the comment should say y is rewritten.

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+    val name = tree.removeAttachment(OriginalName).getOrElse(nme.NO_NAME)
+    if tree.span.isSynthetic && tree.symbol == defn.TypeTest_unapply then
+      tree.qualifier.tpe.underlying.finalResultType match
+      case AppliedType(_, args) => // if tycon.typeSymbol == defn.TypeTestClass
```
documentary? It has already checked for TypeTest.unapply, so the extractor should be a TypeTest[A, B], where the user wrote `case _: B`.

Some combination of redundant code and comments should help me remember later what I was working out, for some definition of "working memory".

som-snytt commented on compiler/src/dotty/tools/dotc/transform/CheckUnused.scala:
```diff
@@ -1,883 +1,867 @@
+
+  type MessageInfo = (UnusedSymbol, SrcPos, String) // string is origin or empty
+
+  def warnings(using Context): Array[MessageInfo] =
```
Yes, I'll undertake that in a separate commit. Probably just a method for each case (params, privates, imports etc). I'll also move the "rewrite" logic out of the way. I did verify that the rewrite correctly edits the imports in the repo.

sjrd replied:
This is much better now. Thank you.

som-snytt commented:
@sjrd I appreciate the review. I have addressed the tweaks and also a first refactor of the `warnings` loop, and also a refactor of `resolveUsage`.

The cache logic was flawed: now it caches the `Precedence` of the result. Otherwise, a subsequent look-up may have a more deeply nested result and doesn't know whether to keep it. (Contexts are ordered but don't bear nesting levels, so the highest precedence candidate wins.)

I will add commits for further accommodation of `compiletime` and `inline`, but do not plan to squash.

You may wish to wait for more tests, or review at your leisure. I'll ping you again if I've exhausted my resources.

som-snytt commented:
I commented that `-Werror` spoils the output of `tests/warn`, but that is not true (at least, as of now):
```
Compilation failed for: tests/warn/i21805.scala
-> following the diagnostics:
 at 0: No warnings can be incurred under -Werror (or -Xfatal-warnings)
 at 7: unused import
```

som-snytt commented:
@sjrd that is probably my limit for this PR

sjrd reviewed (approved):
Looks great! Thanks a lot for reworking that area which definitely deserved more attention.

Do you want to keep the commits as is or squash some of them? In any case, LGTM.

som-snytt commented:
Squashed, fixed conflict in posttyper (!), amended the comment, left the change for `import chaining` as a separate commit for visibility (it's just a tweak but not fully vetted).

I'll expect to follow up, for bugs or known false positives with compiletime ops.

sjrd commented:
There are some test failures in the CI. Can you take a look?

987Nabil commented:
Can this be back ported to 3.3.x?

tgodzik commented:
We'll try to backport it for sure, though this might be quite difficult due to amount of changes. But the plan is to do it.
